lib/spack/spack/cmd/config.py

```diff
@@ -275,12 +290,77 @@ def config_remove(args):
     set_config(args, path, existing, scope)
 
 
+def config_update(args):
```

Comment:
Questions: 

* is there a mechanism to automatically convert the old external format to the new one?
* will Spack be able to use the old and new format configs?
* if a user doesn’t explicitly update, but later on a config file is updated by a Spack command (e.g. `spack external find`), will it be automatically converted to the new format?

Reply:
> will Spack be able to use the old and new format configs? 

No, Spack will be able to use the new format only. If a user has a `packages.yaml` in the old format then an error message is printed:
```console
$ cat ~/.spack/packages.yaml 
packages:
  cmake:
    paths:
      cmake@3.17.3: /usr
$ spack spec cmake
Input spec
--------------------------------
cmake

Concretized
--------------------------------
==> Error: /home/culpo/.spack/packages.yaml:3: the attributes "modules" and "paths" in packages.yaml have been deprecated. Run "spack config update packages" to update your configuration
```
with hints on how to update to the new format. Users can update their configuration using a command, but need to do so explicitly.

> is there a mechanism to automatically convert the old external format to the new one?

Yes, there's a new command `spack config update <section>`. Going back to the example above:
```console
$ spack config update packages
==> File "/home/culpo/.spack/packages.yaml" updated [backup=/home/culpo/.spack/packages.yaml.bkp]

$ cat ~/.spack/packages.yaml.bkp 
packages:
  cmake:
    paths:
      cmake@3.17.3: /usr

$ cat ~/.spack/packages.yaml
packages:
  cmake:
    externals:
    - spec: cmake@3.17.3
      prefix: /usr
```
The update converts the old format to the new one automatically, but users have to give the command explicitly.

> if a user doesn’t explicitly update, but later on a config file is updated by a Spack command (e.g. spack external find), will it be automatically converted to the new format?

If a user has bits in `packages.yaml` that are not compatible with the new format an error will be printed like shown above. As far as Spack is concerned `packages.yaml` is always written in the new format.

Reply:
I think we should try to do this automatically. If a configuration error is detected, we could specifically look for an existing backup file and suggest that the user include their current/preexisting config in a bug report. 

Also: how will this deal with user environments? Can it convert the YAML there? Environments might be one place where the backup approach is problematic. Would it be possible to read the old config into the new data structure and use inferred defaults?

Reply:
> how will this deal with user environments? Can it convert the YAML there?

This is a part I overlooked in the implementation, thanks for spotting it. Having an environment with the old format for `packages.yaml` currently fails with a descriptive error message that ends with a wrong advice to run `spack config update`. I think the best solution here would be to code `spack env update` along the lines of `spack config update`.

I'll implement that and then get back to discuss an automatic update. What you propose is definitely doable but I think we should discuss pros and cons of both approaches before going for one (automatic update requires no effort but can be unexpected, explicit update can be tedious but is done consciously).

Reply:
> The user should get a prompt (Y/n) for whether they want to continue and modify their config files, or abort. That should minimize surprise to the user.

@becker33 If my understanding is correct this is something more similar to what is currently coded than to what @scheibelp proposes, since it still requires user interaction. I think that people might get surprised by something like this:
```console
$ # 1. When an update is needed ask user for confirmation
$ spack install hdf5
==> Your packages.yaml is out of date and will be updated to a new format. Do you want to proceed?
[y/N]
...
```
happening within scripts or CI, since the symptom of it would probably be seeing the process hanging indefinitely. We can use a prompt with a timeout but that might very well be over engineering. What is currently coded requires user interaction AND one more command but has the benefit of failing immediately:
```console
$ # 2. When an update is needed fail and show which command can be used to update
$ spack install hdf5
==> Error: /home/culpo/.spack/packages.yaml:3: the attributes "modules" and "paths" in packages.yaml have been deprecated. 

Run "spack config update packages" to update your configuration
```

My understanding of having automated updates was more something along the line of updating the file and notifying the user without asking for confirmation:
```console
$ # 3. When an update is needed go ahead and notify user of the change
$ spack install hdf5
==> WARNING: your "packages.yaml" is out of date and has been updated to a new format [backup=/some/location]
```
similar considerations can be done for environments. Overall the use cases I can think of are:

A. Interactive use of Spack (here I think 1. is possibly optimal 2. and 3. have different pros and cons)
B. Use of Spack within scripts or CI (here 2. and 3. are better than 1.)

**There's also another caveat** that just came to my mind with this respect i.e.we have `system` scope (`/etc/spack`) and we can't assume to have the permissions to update an outdated `packages.yaml` there :thinking:

Reply:
I'm re-reading this https://github.com/spack/spack/pull/16526#discussion_r425663134 and at a high-level I see as a possible solution to segregate configuration files into subfolders versioned by major Spack versions. To exemplify what I mean, I would turn something like:
```console
/home/culpo/.spack/
├── config.yaml
├── linux
│   ├── compilers.yaml
│   ├── mirrors.yaml
│   └── repos.yaml
├── mirrors.yaml
├── modules.yaml
└── packages.yaml
```
into:
```console
/home/culpo/.spack/
└──v0.15
    ├── config.yaml
    ├── linux
    │   ├── compilers.yaml
    │   ├── mirrors.yaml
    │   └── repos.yaml
    ├── mirrors.yaml
    ├── modules.yaml
    └── packages.yaml
```
and have a single command e.g. `spack config import` to populate the `v0.15` folder by converting previous configuration files if present. This would also be an improvement over the solutions proposed above for people using different major versions of Spack at a time, since multiple configurations can coexist. It is more effort to implement though and if we go for something like that I would probably implement it in a separate PR and rebase this one onto it.

@tgamblin :point_up: What is your opinion on this?

Reply:
I think this is largely resolved now. We agreed on the following strategy:

* Configs are updated as they are read in. This means that users will be able to use old configs with new Spack versions.
* If a config was updated, is is not automatically written out: the user must update it explicitly. This avoids changing user configurations in ways they don't expect (particularly if they were doing any manual editing of these files). This check is done in `config.update_config`.

@becker33 does that seem workable to you?